Use Container instead of Image

miekg opened the pull request:
This clears up the Image == Name confusion. Add some documentation to
the Container, Namespace, Pod and Image functions that makes clearer how they
are used.

The image name is now stored in the X-Kubernetes section of the unit

Closes: #42

Signed-off-by: Miek Gieben <miek@miek.nl>

pires commented on internal/provider/pod.go:
```diff
@@ -406,32 +406,37 @@ func unitPrefix(namespace, podName string) string {
-func Image(name string) string {
-	el := strings.Split(name, separator) // assume well formed
+// Name returns <namespace>.<podname> from a well formed name.
+// Units are named as 'systemk.<namespace>.<podname>.<image> .
```
Shouldn't it be `systemk.<namespace>.<pod>.<container>`? Cause Kubernetes guarantees unique container names per Pod but not images which can repeat in multiple containers of a same Pod.

pires replied:
While I'm backing down on my other comments, it seems we agree this needs to be addressed.

miekg replied:
yep done

pires commented on internal/provider/pod.go:
```diff
@@ -406,32 +406,37 @@ func unitPrefix(namespace, podName string) string {
 
-func Image(name string) string {
-	el := strings.Split(name, separator) // assume well formed
+// Name returns <namespace>.<podname> from a well formed name.
```
What's the definition of a _well-formed name_? I would rather have this take a namespace, pod and container strings than an opaque value that may or may not contain a valid concatenation of those. By doing so, we'd even get rid of string splitting.

pires commented on internal/provider/pod.go:
```diff
@@ -406,32 +406,37 @@ func unitPrefix(namespace, podName string) string {
-func Name(name string) string {
-	el := strings.Split(name, separator) // assume well formed
+// Image returns the <image> from the well formed name. See Name.
+func Image(name string) string {
```
Same comment as above regarding my preference for multiple arguments vs an opaque one and string splitting.

pires commented on internal/provider/unit.go:
```diff
@@ -108,8 +108,8 @@ func (p *p) toContainers(stats map[string]*unit.State) ([]corev1.Container, []co
 		s := stats[k]
 		u, _ := unit.NewFile(s.UnitData)
 		container := v1.Container{
```
String splitting and validation could happen here and it would only be done once instead of per call to `Name` and `Image`.

miekg replied:
this unit name splitting is done in various places, so I don't understand how this would help with that?

pires replied:
Maybe the context of the PR is not enough and it requires me to go through the code and see where this is called. So I'll take a step back on this one.

miekg commented:
[ Quoting <notifications@github.com> in "Re: [virtual-kubelet/systemk] Use N..." ]
>@pires commented on this pull request.
>
>━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
>
>In internal/provider/pod.go:
>
>> @@ -406,32 +406,37 @@ func unitPrefix(namespace, podName string) string {
>        return prefix + separator + namespace + separator + podName
> }
>
>-func Image(name string) string {
>-       el := strings.Split(name, separator) // assume well formed
>+// Name returns <namespace>.<podname> from a well formed name.
>+// Units are named as 'systemk.<namespace>.<podname>.<image> .
>
>Shouldn't it be systemk.<namespace>.<pod>.<container>? Cause Kubernetes
>guarantees unique container names per Pod but not images which can repeat in
>multiple containers of a same Pod.

Good point, and this was *also* something the confused me back then. And now the comment
makes sense "we don't store the image name" because this is container. Image then need
to go into X-Kubernetes.

>━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
>
>In internal/provider/pod.go:
>
>> @@ -406,32 +406,37 @@ func unitPrefix(namespace, podName string) string {
>        return prefix + separator + namespace + separator + podName
> }
>
>-func Image(name string) string {
>-       el := strings.Split(name, separator) // assume well formed
>+// Name returns <namespace>.<podname> from a well formed name.
>
>What's the definition of a well-formed name? I would rather have this take a
>namespace, pod and container strings than an opaque value that may or may not
>contain a valid concatenation of those. By doing so, we'd even get rid of
>string splitting.

how? We get this strings from systemd.

pires commented:
On your last question, here's [my suggestion](https://github.com/virtual-kubelet/systemk/pull/77#discussion_r566025081).

pires reviewed (approved):
LGTM!
